File: Java/Eta/Applications/Examples/com/thomsonreuters/upa/valueadd/examples/common/MarketPriceRefresh.java
package com.thomsonreuters.upa.valueadd.examples.common;

import com.thomsonreuters.upa.codec.CodecFactory;
import com.thomsonreuters.upa.codec.CodecReturnCodes;
import com.thomsonreuters.upa.codec.DataDictionary;
import com.thomsonreuters.upa.codec.DataTypes;
import com.thomsonreuters.upa.codec.DictionaryEntry;
import com.thomsonreuters.upa.codec.EncodeIterator;
import com.thomsonreuters.upa.codec.Msg;
import com.thomsonreuters.upa.codec.MsgClasses;
import com.thomsonreuters.upa.codec.Qos;
import com.thomsonreuters.upa.codec.RefreshMsg;
import com.thomsonreuters.upa.codec.State;
import com.thomsonreuters.upa.rdm.DomainTypes;
import com.thomsonreuters.upa.rdm.InstrumentNameTypes;

/**
 * The market price refresh message. Used by an OMM Consumer, OMM
 * Non-Interactive Provider and OMM Interactive provider to encode/decode a
 * market price refresh message.
 */
public class MarketPriceRefresh extends MarketPriceResponseBase
{
    private State state;
    private Qos qos;

    // used for encoding
    private RefreshMsg refreshMsg = (RefreshMsg)CodecFactory.createMsg();

    public MarketPriceRefresh()
    {
        state = CodecFactory.createState();
        qos = CodecFactory.createQos();
    }

    /**
     * Clears the current contents of this object and prepares it for re-use.
     */
    public void clear()
    {
        super.clear();
        state.clear();
        qos.clear();
        flags = 0;
    }

  
    /**
     * state of the market price.
     * 
     * @return {@link State}
     */
    public State state()
    {
        return state;
    }

    /**
     * qos value of the market price.
     * 
     * @return {@link Qos}
     */
    public Qos qos()
    {
        return qos;
    }

    /**
     * Checks the presence of refresh complete flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkRefreshComplete()
    {
        return (flags & MarketPriceRefreshFlags.REFRESH_COMPLETE) != 0;
    }

    /**
     * Applies refresh complete flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyRefreshComplete()
    {
        flags |= MarketPriceRefreshFlags.REFRESH_COMPLETE;
    }
    
    /**
     * Checks the presence of clear cache flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkClearCache()
    {
        return (flags & MarketPriceRefreshFlags.CLEAR_CACHE) != 0;
    }

    /**
     * Applies clear cache flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyClearCache()
    {
        flags |= MarketPriceRefreshFlags.CLEAR_CACHE;
    }

    /**
     * Checks the presence of solicited flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkSolicited()
    {
        return (flags & MarketPriceRefreshFlags.SOLICITED) != 0;
    }

    /**
     * Applies solicited flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applySolicited()
    {
        flags |= MarketPriceRefreshFlags.SOLICITED;
    }

    /**
     * Checks the presence of Qos flag.
     * 
     * This flag can also be bulk-get by {@link #flags()}
     * 
     * @return true - if exists; false if does not exist.
     */
    public boolean checkHasQos()
    {
        return (flags & MarketPriceRefreshFlags.HAS_QOS) != 0;
    }

    /**
     * Applies qos flag.
     * 
     * This flag can also be bulk-set by {@link #flags(int)}
     */
    public void applyHasQos()
    {
        flags |= MarketPriceRefreshFlags.HAS_QOS;
    }

    public Msg encodeMsg()
    {
        refreshMsg.clear();

        refreshMsg.msgClass(MsgClasses.REFRESH);
        refreshMsg.streamId(streamId());
        refreshMsg.applyHasMsgKey();

        if (checkPrivateStream())
        {
            refreshMsg.applyPrivateStream();
        }

        if (checkRefreshComplete())
        {
            refreshMsg.applyRefreshComplete();
        }
        
        if (checkClearCache())
        {
            refreshMsg.applyClearCache();
        }
        
        if (checkSolicited())
        {
            refreshMsg.applySolicited();
        }

        if (checkPrivateStream())
        {
            refreshMsg.applyPrivateStream();
        }

        if (checkHasServiceId())
        {
            refreshMsg.msgKey().applyHasServiceId();
            refreshMsg.msgKey().serviceId(serviceId());
        }

        // Itemname
        refreshMsg.msgKey().name(itemName());
        refreshMsg.msgKey().applyHasName();
        refreshMsg.msgKey().nameType(InstrumentNameTypes.RIC);
        refreshMsg.msgKey().applyHasNameType();

        // State
        refreshMsg.state().dataState(state().dataState());
        refreshMsg.state().streamState(state().streamState());
        refreshMsg.state().code(state().code());
        refreshMsg.state().text(state().text());

        if (checkHasQos())
        {
            refreshMsg.applyHasQos();
            refreshMsg.qos().dynamic(qos().isDynamic());
            refreshMsg.qos().rate(qos().rate());
            refreshMsg.qos().timeliness(qos().timeliness());
            refreshMsg.qos().rateInfo(qos().rateInfo());
            refreshMsg.qos().timeInfo(qos().timeInfo());
        }

        refreshMsg.domainType(DomainTypes.MARKET_PRICE);
        refreshMsg.containerType(DataTypes.FIELD_LIST);

        return refreshMsg;
    }

    public String toString()
    {
        StringBuilder toStringBuilder = super.buildStringBuffer();
        toStringBuilder.insert(0, "MarketPriceRefresh: ");

        toStringBuilder.append(tabChar);
        toStringBuilder.append("state: ");
        toStringBuilder.append(state());
        toStringBuilder.append(eolChar);

        toStringBuilder.append(tabChar);
        toStringBuilder.append("solicited: ");
        toStringBuilder.append(checkSolicited());
        toStringBuilder.append(eolChar);

        toStringBuilder.append(tabChar);
        toStringBuilder.append("refreshComplete: ");
        toStringBuilder.append(checkRefreshComplete());
        toStringBuilder.append(eolChar);

        if (checkHasServiceId())
        {
            toStringBuilder.append(tabChar);
            toStringBuilder.append("serviceId: ");
            toStringBuilder.append(serviceId());
            toStringBuilder.append(eolChar);
        }
        if (checkHasQos())
        {
            toStringBuilder.append(tabChar);
            toStringBuilder.append("qos: ");
            toStringBuilder.append(refreshMsg.qos().toString());
            toStringBuilder.append(eolChar);
        }

        return toStringBuilder.toString();
    }

    protected int encodeRefreshFields(EncodeIterator encodeIter, DataDictionary dictionary)
    {
        MarketPriceItem mpItem = marketPriceItem();
        DictionaryEntry dictionaryEntry;
        int ret = CodecReturnCodes.SUCCESS;
        
        // RDNDISPLAY
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketPriceItem.RDNDISPLAY_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketPriceItem.RDNDISPLAY_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            tempUInt.value(mpItem.RDNDISPLAY);
            if ((ret = fieldEntry.encode(encodeIter, tempUInt)) < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }
        // RDN_EXCHID
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketPriceItem.RDN_EXCHID_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketPriceItem.RDN_EXCHID_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            com.thomsonreuters.upa.codec.Enum enumValue = CodecFactory.createEnum();
            enumValue.value(mpItem.RDN_EXCHID);
            if ((ret = fieldEntry.encode(encodeIter, enumValue)) < CodecReturnCodes.SUCCESS)
            {
                return ret;
            }
        }
        
        // DIVPAYDATE
        fieldEntry.clear();
        dictionaryEntry = dictionary.entry(MarketPriceItem.DIVPAYDATE_FID);
        if (dictionaryEntry != null)
        {
            fieldEntry.fieldId(MarketPriceItem.DIVPAYDATE_FID);
            fieldEntry.dataType(dictionaryEntry.rwfType());
            return fieldEntry.encode(encodeIter, mpItem.DIVPAYDATE);
        }

        return ret;
    }
    
    @Override
    public int domainType()
    {
        return DomainTypes.MARKET_PRICE;
    }
}

File: CSharp/Eta/Applications/Examples/EtaAppCommon/Rdm/item/marketprice/MarketPriceRefresh.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2022-2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Codec;
using LSEG.Eta.Rdm;
using System.Diagnostics;
using System.Text;
using Enum = LSEG.Eta.Codec.Enum;

namespace LSEG.Eta.Example.Common
{
    public class MarketPriceRefresh : MarketPriceBase
    {
        private Msg m_RefreshMsg = new Msg();
        private Qos m_Qos = new Qos();

        private UInt tmpUInt = new UInt();

        public override int StreamId { get => m_RefreshMsg.StreamId; set { m_RefreshMsg.StreamId = value; } }
        public override int MsgClass { get => m_RefreshMsg.MsgClass; }
        public override int DomainType { get => m_RefreshMsg.DomainType; }
        public bool Solicited
        {
            get => (Flags & ItemDomainCommonFlags.SOLICITED) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.SOLICITED;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.SOLICITED;
            }
        }
        public bool RefreshComplete
        {
            get => (Flags & ItemDomainCommonFlags.REFRESH_COMPLETE) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.REFRESH_COMPLETE;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.REFRESH_COMPLETE;
            }
        }
        public bool ClearCache 
        { 
            get => (Flags & ItemDomainCommonFlags.CLEAR_CACHE) != 0; 
            set 
            { 
                if (value) { Flags |= ItemDomainCommonFlags.CLEAR_CACHE; } 
                else Flags &= ~ItemDomainCommonFlags.CLEAR_CACHE; 
            } 
        }
        public bool HasQos
        {
            get => (Flags & ItemDomainCommonFlags.HAS_QOS) != 0;
            set
            {
                if (value)
                {
                    Flags |= ItemDomainCommonFlags.HAS_QOS;
                }
                else
                    Flags &= ~ItemDomainCommonFlags.HAS_QOS;
            }
        }
        public State State
        {
            get => m_RefreshMsg.State;
            set
            {
                Debug.Assert(value != null);
                m_RefreshMsg.State.StreamState(value.StreamState());
                m_RefreshMsg.State.DataState(value.DataState());
                m_RefreshMsg.State.Code(value.Code());
                m_RefreshMsg.State.Text(value.Text());
            }
        }
        public Qos Qos
        {
            get => m_Qos; 
            set 
            {
                Debug.Assert(value != null);
                value.Copy(m_Qos);
            }
        }
        public long Filter { get => m_RefreshMsg.MsgKey.Filter; set { m_RefreshMsg.MsgKey.Filter = value; } } 
        
        public MarketPriceRefresh()
        {
            Clear();
        }
  
        public override void Clear()
        {
            m_Qos.Clear();
            Flags = 0;
            m_RefreshMsg.Clear();
            m_RefreshMsg.DomainType = (int)Rdm.DomainType.MARKET_PRICE;
            m_RefreshMsg.MsgClass = MsgClasses.REFRESH;
        }

        public override Msg EncodeMsg()
        {
            if (PrivateStream)
            {
                m_RefreshMsg.ApplyPrivateStream();
            }

            if (RefreshComplete)
            {
                m_RefreshMsg.ApplyRefreshComplete();
            }

            if (ClearCache)
            {
                m_RefreshMsg.ApplyClearCache();
            }

            if (Solicited)
            {
                m_RefreshMsg.ApplySolicited();
            }

            m_RefreshMsg.ApplyHasMsgKey();

            if (HasServiceId)
            {
                m_RefreshMsg.MsgKey.ApplyHasServiceId();
                m_RefreshMsg.MsgKey.ServiceId = ServiceId;
            }

            // Itemname
            m_RefreshMsg.MsgKey.Name = ItemName;
            m_RefreshMsg.MsgKey.ApplyHasName();
            m_RefreshMsg.MsgKey.NameType = InstrumentNameTypes.RIC;
            m_RefreshMsg.MsgKey.ApplyHasNameType();

            // State
            State.Copy(m_RefreshMsg.State);

            if (HasQos)
            {
                m_RefreshMsg.ApplyHasQos();
                Qos.Copy(m_RefreshMsg.Qos);
            }
            m_RefreshMsg.ContainerType = DataTypes.FIELD_LIST;

            return m_RefreshMsg;
        }

        public override string ToString()
        {
            StringBuilder toStringBuilder = PrepareStringBuilder();
            toStringBuilder.Insert(0, "MarketPriceRefresh: ");

            toStringBuilder.Append(tab);
            toStringBuilder.Append("state: ");
            toStringBuilder.Append(State);
            toStringBuilder.AppendLine();

            toStringBuilder.Append(tab);
            toStringBuilder.Append("solicited: ");
            toStringBuilder.Append(Solicited);
            toStringBuilder.AppendLine();

            toStringBuilder.Append(tab);
            toStringBuilder.Append("refreshComplete: ");
            toStringBuilder.Append(RefreshComplete);
            toStringBuilder.AppendLine();

            if (HasServiceId)
            {
                toStringBuilder.Append(tab);
                toStringBuilder.Append("serviceId: ");
                toStringBuilder.Append(ServiceId);
                toStringBuilder.AppendLine();
            }
            if (HasQos)
            {
                toStringBuilder.Append(tab);
                toStringBuilder.Append("qos: ");
                toStringBuilder.Append(m_RefreshMsg.Qos.ToString());
                toStringBuilder.AppendLine();
            }

            return toStringBuilder.ToString();
        }

        protected override CodecReturnCode EncodeRefreshFields(EncodeIterator encodeIter, DataDictionary dictionary)
        {
            MarketPriceItem mpItem = MarketPriceItem!;
            IDictionaryEntry dictionaryEntry;
            CodecReturnCode ret = CodecReturnCode.SUCCESS;

            // RDNDISPLAY
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketPriceItem.RDNDISPLAY_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketPriceItem.RDNDISPLAY_FID;
                fieldEntry.DataType = dictionaryEntry.GetRwfType();
                tempUInt.Value(mpItem.RDNDISPLAY);
                if ((ret = fieldEntry.Encode(encodeIter, tempUInt)) < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }
            // RDN_EXCHID
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketPriceItem.RDN_EXCHID_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketPriceItem.RDN_EXCHID_FID;
                fieldEntry.DataType = dictionaryEntry.GetRwfType();
                Enum enumValue = new Enum();
                enumValue.Value(mpItem.RDN_EXCHID);
                if ((ret = fieldEntry.Encode(encodeIter, enumValue)) < CodecReturnCode.SUCCESS)
                {
                    return ret;
                }
            }

            // DIVPAYDATE
            fieldEntry.Clear();
            dictionaryEntry = dictionary.Entry(MarketPriceItem.DIVPAYDATE_FID);
            if (dictionaryEntry != null)
            {
                fieldEntry.FieldId = MarketPriceItem.DIVPAYDATE_FID;
                fieldEntry.DataType = dictionaryEntry.GetRwfType();
                return fieldEntry.Encode(encodeIter, mpItem.DIVPAYDATE);
            }

            return ret;
        }
    }

    
}
